core/state_transition.go

```diff
@@ -682,53 +785,62 @@ func (st *StateTransition) innerTransitionDb() (*ExecutionResult, error) {
-		// otherwise we add to the native balance
-		if st.gasFromSoul && st.evm.ChainConfig().IsOptimism() && !st.evm.ChainConfig().Optimism.IsSoulBackedByNative {
-		} else {
-			st.state.AddBalance(st.evm.Context.Coinbase, fee, tracing.BalanceIncreaseRewardTransactionFee)
+		st.tipFee = fee.Clone()
 
-			// add the coinbase to the witness iff the fee is greater than 0
-			if rules.IsEIP4762 && fee.Sign() != 0 {
-				st.evm.AccessEvents.AddAccount(st.evm.Context.Coinbase, true)
-			}
+		fee = st.collectableNativeBalance(fee)
+		st.state.AddBalance(st.evm.Context.Coinbase, fee, tracing.BalanceIncreaseRewardTransactionFee)
+
+		// add the coinbase to the witness iff the fee is greater than 0
+		if rules.IsEIP4762 && fee.Sign() != 0 {
+			st.evm.AccessEvents.AddAccount(st.evm.Context.Coinbase, true)
 		}
 	}
 
 	// Check that we are post bedrock to enable op-geth to be able to create pseudo pre-bedrock blocks (these are pre-bedrock, but don't follow l2 geth rules)
 	// Note optimismConfig will not be nil if rules.IsOptimismBedrock is true
 	if optimismConfig := st.evm.ChainConfig().Optimism; optimismConfig != nil && rules.IsOptimismBedrock && !st.msg.IsDepositTx {
 		gasCost := new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), st.evm.Context.BaseFee)
+
+		if st.evm.ChainConfig().IsCancun(st.evm.Context.BlockNumber, st.evm.Context.Time) {
+			gasCost.Add(gasCost, new(big.Int).Mul(new(big.Int).SetUint64(st.blobGasUsed()), st.evm.Context.BlobBaseFee))
+		}
+
 		amtU256, overflow := uint256.FromBig(gasCost)
 		if overflow {
 			return nil, fmt.Errorf("optimism gas cost overflows U256: %d", gasCost)
 		}
-		// we burn the token if gas is from SoulGasToken which is not backed by native;
-		// otherwise we add to the native balance
-		if st.gasFromSoul && st.evm.ChainConfig().IsOptimism() && !st.evm.ChainConfig().Optimism.IsSoulBackedByNative {
-		} else {
-			st.state.AddBalance(params.OptimismBaseFeeRecipient, amtU256, tracing.BalanceIncreaseRewardTransactionFee)
+		if shouldCheckGasFormula {
+			st.baseFee = amtU256.Clone()
 		}
+
+		amtU256 = st.collectableNativeBalance(amtU256)
+		st.state.AddBalance(params.OptimismBaseFeeRecipient, amtU256, tracing.BalanceIncreaseRewardTransactionFee)
 		if l1Cost := st.evm.Context.L1CostFunc(st.msg.RollupCostData, st.evm.Context.Time); l1Cost != nil {
 			amtU256, overflow = uint256.FromBig(l1Cost)
 			if overflow {
 				return nil, fmt.Errorf("optimism l1 cost overflows U256: %d", l1Cost)
 			}
-			// we burn the token if gas is from SoulGasToken which is not backed by native;
-			// otherwise we add to the native balance
-			if st.gasFromSoul && st.evm.ChainConfig().IsOptimism() && !st.evm.ChainConfig().Optimism.IsSoulBackedByNative {
-			} else {
-				st.state.AddBalance(params.OptimismL1FeeRecipient, amtU256, tracing.BalanceIncreaseRewardTransactionFee)
+
+			if shouldCheckGasFormula {
+				st.l1Fee = amtU256.Clone()
+				if err := st.checkGasFormula(); err != nil {
```

Comment:
should call the check to be after `if l1Cost := ...`?

Reply:
Done in 3d5c5a91c6318123f89b64745ab2f0676eed5132.